File: backend/app/schemas/analytics.py
import uuid
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class DriverAnalyticsResponse(BaseModel):
    driver_id: str
    safety_score: float = Field(..., ge=0.0, le=100.0)
    risk_level: str
    harsh_braking_events: int = 0
    rapid_acceleration_events: int = 0
    speeding_events: int = 0
    excessive_idle_events: int = 0
    total_observations: int = 0
    summary: str

    model_config = ConfigDict(from_attributes=True)


class SafetyScoreResponse(BaseModel):
    driver_id: str
    safety_score: float = Field(..., ge=0.0, le=100.0)
    risk_level: str
    summary: str

    model_config = ConfigDict(from_attributes=True)


class AnomalyRecordResponse(BaseModel):
    vehicle_id: str
    timestamp: str
    metric_name: str
    metric_value: float
    expected_range: str
    anomaly_score: float
    anomaly_reason: str
    subsystem: str
    status: str


class AnomalyListResponse(BaseModel):
    vehicle_id: Optional[str] = None
    anomalies: List[AnomalyRecordResponse] = []


class VehicleHealthResponse(BaseModel):
    vehicle_id: str
    health_score: float = Field(..., ge=0.0, le=100.0)
    status: str
    risk_factors: List[str] = []
    summary: str


class MaintenancePredictionResponse(BaseModel):
    vehicle_id: str
    risk_score: float = Field(..., ge=0.0, le=100.0)
    risk_level: str
    recommendation: str
    estimated_rul_km: int
    contributing_factors: List[Dict[str, Any]] = []

File: frontend/src/api/analytics.ts
import { apiClient } from "./client";

export interface DriverAnalyticsResponse {
  driver_id: string;
  safety_score: number;
  risk_level: string;
  harsh_braking_events: number;
  rapid_acceleration_events: number;
  speeding_events: number;
  excessive_idle_events: number;
  total_observations: number;
  summary: string;
}

export interface VehicleHealthResponse {
  vehicle_id: string;
  health_score: number;
  status: "GOOD" | "WARNING" | "CRITICAL";
  risk_factors: string[];
  summary: string;
}

export interface ContributingFactor {
  factor: string;
  weight: number;
}

export interface MaintenancePredictionResponse {
  vehicle_id: string;
  risk_score: number;
  risk_level: string;
  recommendation: string;
  estimated_rul_km: number;
  contributing_factors: ContributingFactor[];
}

export interface AnomalyItem {
  id?: string;
  vehicle_id?: string;
  timestamp?: string;
  metric_name: string;
  metric_value: number;
  expected_range: string;
  anomaly_score: number;
  anomaly_reason: string;
  subsystem: string;
  status: string;
}

export interface AnomalyListResponse {
  vehicle_id?: string | null;
  anomalies: AnomalyItem[];
}

export const analyticsApi = {
  getDriverAnalytics: (driverId: string) =>
    apiClient<DriverAnalyticsResponse>(`/analytics/drivers/${driverId}`),

  getVehicleHealth: (vehicleId: string) =>
    apiClient<VehicleHealthResponse>(`/analytics/vehicles/${vehicleId}/health`),

  getMaintenancePrediction: (vehicleId: string) =>
    apiClient<MaintenancePredictionResponse>(`/analytics/vehicles/${vehicleId}/maintenance`),

  getAnomalies: (params?: { vehicle_id?: string; limit?: number }) => {
    const searchParams = new URLSearchParams();
    if (params?.vehicle_id) searchParams.append("vehicle_id", params.vehicle_id);
    if (params?.limit) searchParams.append("limit", params.limit.toString());
    const query = searchParams.toString();
    return apiClient<AnomalyListResponse>(`/analytics/telemetry/anomalies${query ? `?${query}` : ""}`);
  },
};
